File: Basic/5.cpp
#define _USE_MATH_DEFINES
#include <cmath>

const double ONE_MINUTE_DEG = 360.0 / 60;
const double ONE_HOUR_DEG = 360.0 / 12;
// Доп. смещение часовой стрелки от минут внутри часа
const double HOUR_1_MINUTE_OFFSET_DEG = ONE_HOUR_DEG / 60;

int main() {
    int hours = 14, minutes = 30;

    double minute_angle = ONE_MINUTE_DEG * minutes;
    double hour_angle = ONE_HOUR_DEG * (hours % 12) + HOUR_1_MINUTE_OFFSET_DEG * minutes;

    double angle = std::abs(minute_angle - hour_angle);
    if (angle >= 180)
        angle = 360 - angle;

    return 0;
}

File: Basic/5.hs
main = do
    hours <- input "Часы" :: IO Int
    minutes <- input "Минуты" :: IO Int
    putStrLn (show (angle hours minutes) ++ " градусов")

input :: Read b => [Char] -> IO b
input prompt = do
    putStr $ prompt ++ ": "
    read <$> getLine

angle :: Int -> Int -> Double
angle hours minutes = if angle >= 180 then 360 - angle else angle
    where
        oneMinuteDeg = 360 / 60
        oneHourDeg = 360 / 12
        hourMinuteOffsetDeg = oneHourDeg / 60
        minuteAngle = oneMinuteDeg * (fromIntegral minutes :: Double)
        hourAngle = oneHourDeg * fromIntegral (hours `mod` 12) + hourMinuteOffsetDeg * fromIntegral minutes
        angle = abs $ minuteAngle - hourAngle
